ddtrace/settings/integration.py

```diff
@@ -1,62 +1,239 @@
 from copy import deepcopy
 
-from ..utils.attrdict import AttrDict
+from ..vendor import attr
 from ..utils.formats import asbool, get_env
 from .http import HttpConfig
 from .hooks import Hooks
 
 
-class IntegrationConfig(AttrDict):
+@attr.s
+class Setting(object):
+    """Setting represents a configuration setting which is defined by the
+    library and optionally overridden by the user.
     """
-    Integration specific configuration object.
+    @attr.s
+    class Undefined(object):
+        value = attr.ib(default=None)
 
-    This is what you will get when you do::
+    Undefined = Undefined()
+
+    @attr.s
+    class Defined(object):
+        value = attr.ib()
+
+    class SettingKeyError(Exception):
+        pass
+
+    _value = attr.ib(default=Undefined, validator=attr.validators.instance_of(Defined))
+    _default = attr.ib(init=False, default=Undefined)
+
+    def __attrs_post_init__(self):
+        self._default = self._value
+
+    def set(self, value):
+        assert value is not self.Undefined
+        assert not isinstance(value, Setting)
+        self._value = value
+        return value
+
+    def get(self):
+        if self._value is self.Undefined:
+            raise self.SettingKeyError()
+        return self._value.value
+
+    def reset(self):
+        self._value = self._default
+
+    def isdefined(self):
+        return self._value is not self.Undefined
+
+    def check(self, _type):
+        return isinstance(self._value, _type)
+
+
+class IntegrationConfig(object):
+    """
+    All internal methods/properties must be private (prefixed with _)
+
+    - Differentiation between library and user defined values
+    - Support for defaults
+    - User API is limited to reading/writing only settings that exist
+    - Awareness of whether a setting has been overridden by a user
+
+    Library API::
 
         from ddtrace import config
+        config._add("redis", dict(
+            service="redis",
+        ))
+        config.redis._get("service")
+
 
-        # This is an `IntegrationConfig`
-        config.flask
+    User API::
+
+        from ddtrace import config
+        config.redis.service = "test"
+        config.redis.serivce = "test"  # throws exception!!
 
-        # `IntegrationConfig` supports both attribute and item accessors
-        config.flask['service_name'] = 'my-service-name'
-        config.flask.service_name = 'my-service-name'
     """
-    def __init__(self, global_config, name, *args, **kwargs):
-        """
-        :param global_config:
-        :type global_config: Config
-        :param args:
-        :param kwargs:
+
+    class UserDefined(Setting.Defined):
+        pass
+
+    class LibDefined(Setting.Defined):
+        pass
+
+    class IntegrationConfigAttributeError(AttributeError):
+        pass
+
+    class IntegrationConfigKeyError(KeyError):
+        pass
+
+    @attr.s
+    class Dec(object):
```

Comment:
pardon my English, I don't get why this is named `Dec`?

Reply:
I was going for "Declaration" maybe.. lol this was certainly not intended to be the real thing... I don't really like it either 😅